### ISSUE TITLE
Namespace for modules

### ISSUE THREAD

**Issue by certik** [lfortran/fortran_proposals#1]
So you can require module data to be referenced like module.var:
```fortran
use utils, only
...
call utils%savetxt()
```
One issue with just `use utils` is that it still pollutes your local namespace with `savetxt`, so that's the reason for `use utils, only` (perhaps one could do `use utils, only:`).

Another alternative is `use namespace utils` (suggested by [Milan Curcic](https://twitter.com/realmilancurcic/status/1184269876543799296)). Or perhaps `use, namespace :: utils`, to by compatible with the existing syntax like `use, intrinsic :: iso_fortran_env`.

Initially discussed at https://twitter.com/OndrejCertik/status/1184245450976575488.

**Comment by certik** [comment]
The paper is submitted at https://j3-fortran.org/doc/year/19/19-246.txt. I will try to get feedback on it from the committee.

**Comment by septcolor** [comment]
+1 for this proposal because, in addition to that many other languages
provide this feature, I often create a dummy derived type (having a lot of
NOPASS module variables) to mimic this behavior (which is very tedious).

Indeed, this seems to be almost a "frequently asked request (FAR)" (not FAQ :-)
http://fortranwiki.org/fortran/show/Extension+to+USE+statement

**Comment by cmacmackin** [comment]
Agree, this seems like a good feature.

**Comment by certik** [comment]
So the committee unfortunately did not consider this proposal at the meeting last week. I was hoping to get at least a general "yes" or "no" direction from the committee. But I talked to a few members directly, and I did not hear any "this is a bad idea", or "this should not be done". 

Here is my plan:

1. Write a high quality full proposal for this. This means more use cases, we have to brainstorm any possible issues, and some members of the committee like to structure any new features as "requirements", "specification" and "syntax", so we should incorporate all that.

2. I will ask on the j3 mailinglist once a more complete proposal is ready to get some feedback

3. I will ensure that this gets to the committee program for next meeting to be official considered for the 202y standard.

**Comment by septcolor** [comment]
I fully agree with plan (1-3), but I have one concern (which may apply to other proposals as well). Though I may be too nervous, I will explain it anyway...

Specifically, even if the proposal describes several use cases + comparison with other languages + possible new syntax, I'm afraid that many of the committee people may not "intuitively" get the utility of the proposal _unless_ they have some real experience of using it in other languages and experiencing its utility. I felt this strongly when I read the discussions and responses about generics on comp.lang.fortran (much ago). There, the committee wanted to know the use cases of generics (apart from swap()), and it seems that only statistics (from the user survey) became the reason for considering generics as a candidate for revision. So I have some concern that the same thing might happen for other proposals (i.e., they do not "feel" good or bad about a new proposal and so possibly not willing to give either positive or negative responses).

But my concern above may be too nervous... (Still, I'm wondering if there is some means that the committee people would "feel" its utility via experience :)

**Comment by certik** [comment]
@septcolor that's a valid concern. My plan to overcome this is to use this github repository to get support for proposals. If the Fortran community puts thumbs up for a proposal, or otherwise expresses strong interest, then relatively to other ideas proposed here it will be obvious which proposals are high priority, and which are low priority. And then it will be easier to argue for a particular proposal at the committee.

**Comment by FortranFan** [comment]
Fyi a new thread at comp.lang.fortran inquiring of a missing feature in current Fortran which can be met by the proposal here:
https://groups.google.com/d/msg/comp.lang.fortran/nYdFG0PCTWo/rjjLlAteDgAJ

**Comment by certik** [comment]
Thanks @FortranFan. This is one of the popular issues in here and I am going to insist for the committee to consider it at the next meeting. However, I welcome any help with improving the current proposal (linked above).

**Comment by FortranFan** [comment]
@certik wrote:

> ..However, I welcome any help with improving the current proposal (linked above).

@certik and others interested in this, if you haven't already please take a note of comments online, especially by Ian Harvey at comp.lang.fortran such as this [**one**](https://groups.google.com/d/msg/comp.lang.fortran/nYdFG0PCTWo/u0bLHpkcDwAJ): 

Look at the last 3 questions in that comment: addressing these questions suitably will strengthen the proposal.  Some aspects that come to my mind with these questions include:

1. USE statements can include unnamed operators (+,-*,/, etc.), assignment(=), etc.  What is the impact of this proposal on their semantics?
2. What about named entities which are operators in USE statements e.g., operator(.cross_product.)?  Does ```x .cross_product. y``` become ```x foo::operator(.cross_product.) y``` a la C++?
3. What about generic interfaces and their extension via USE statements?
4. Current standard has scoping unit semantics with MODULE subroutines/functions (e.g., for use in SUBMODULEs), INTERFACEs. BLOCK constructs, etc. that can be impacted by USE statements.  What happens with them?

**Comment by certik** [comment]
@FortranFan thanks a lot for pointing these potential issues out. Let's address them.

**Comment by septcolor** [comment]
As for why a more explicit namespace control is useful (and so commonly used by other languages), I think there may be at least several reasons:

- Minimize the degree of symbol/name contamination in the current scope
- Enhance the readability of a code via explicit qualifiers (no need to search for
"where this variable comes from")
- DRY (don't repeat yourself) https://en.wikipedia.org/wiki/Don%27t_repeat_yourself
- Potentially benefit from a convenient feature for code completion by modern editors and IDEs (e.g. "Intellisense" https://docs.microsoft.com/en-us/visualstudio/ide/using-intellisense?view=vs-2019)

---
Reference a module like a global "object"

In object-oriented (OO) programing, we know that it is very convenient to have the ability to refer to components and TBP via the "%" symbol. What we want is close to this ability for a module, such that we can reference an entire module as a global object (like "singleton"). We could mimic its behavior by creating a dummy type that provides aliases to all module variables and contained routines (e.g., via pointers and TBP with NOPASS), but this approach is sometimes rather tedious and may not be straightforward for 3rd-party libraries. I think this kind of extra work becomes unnecessary when a qualifier-based module access is provided at the language level.

---
About DRY vs WET

The DRY Wiki page mentions WET as an anti-pattern (i.e., not good practice), e.g. "write everything twice" and "we enjoy typing". To import a routine with a long name, I have to type it twice in current Fortran (i.e., once for a USE statement and once for actuall CALL statement). I often find this frustrating when I need to import many routines with long names. Sometimes, it is suggested to use the rename facility of USE (e.g., `use foo_mod, only: shortname => long_routine_name`), but I feel this approach is often not ideal because of the need to consider yet another name for each imported routine and also because the reader needs to search for their local definition at the top of the current scope.

---
Matplotlib may be a good "use case" of an explicit qualifier (mpl, plt, etc)

I think a good "real-world" use case may be Python/matplotlib, which commonly use the qualifier "mpl", "plt", and so on. Because Matplotlib contains tons of names/symbols inside, it is clearly not good practice to import all symbols at the same time (via star import). On the other hand, if Python had only Fortran-like import facility, it would have to type every routine names twice, which is very tedious and lengthy in practice.

In addition to the case of Lapack and redundantly named routines (as described in https://j3-fortran.org/doc/year/19/19-246.txt),

**Comment by gronki** [comment]
I think calling DRY good and WET bad is rather arbitrary and not a good
argument. More explicit programming can help avoid some mistakes. In python
I prefer importing explicitly all names that I use in the code I write. If
I feel lazy or only do fast prototyping I import the whole module.

I don't see the discussed feature as necessary but that's because I
probably will not use it. But for those who prefer that programming style
it might be good to have that option.

Dominik

pon., 18 lis 2019, 18:19 użytkownik septcolor <notifications@github.com>
napisał:

**Comment by septcolor** [comment]
@gronk Yes, programming styles depend on people, and I do not claim that one style is more "correct" over the other. For example, I do not strictly follow [PEP8](https://www.python.org/dev/peps/pep-0008/) in my (personal) Fortran codes, despite some people recommend it as a "good" style even for other languages.

On the other hand, one fact is that current Fortran does not allow one to use a module name as a qualifier (in contrast to all other languages I use), which limits us to stick to one approach over the other (i.e., write each imported name redundantly in a local scope).

RE whether to use qualifiers or not, I think it also depends on cases. For example, some people insist using "std::" in C++ always, but I often import specific entities locally for brevity (unless ambiguity occurs). Here, the choice also depends on coding styles and approaches to some extent, I think.

**Comment by qolin1** [comment]
Gets my vote FWIW. I have wanted this for some time, without knowing its name was "namespaces".

**Comment by certik** [comment]
@qolin1 I just called it "namespaces" for a lack of a better name. We will keep pushing for this at our next committee meeting. I think this is a popular feature.

**Comment by klausler** [comment]
> @qolin1 I just called it "namespaces" for a lack of a better name. We will keep pushing for this at our next committee meeting. I think this is a popular feature.

But we need to prioritize.  This change is a convenience, not a fix or an enabling feature.

**Comment by gronki** [comment]
I would personally deem this low priority too.

pon., 23 gru 2019, 18:00 użytkownik Peter Klausler <notifications@github.com>
napisał:

**Comment by certik** [comment]
@klausler, @gronki make sure you put thumbs up on issues that you find high priority, and let's collaborate on the top priority issues. I just created #122 for that purpose, let's discuss the list of top priority issues there.

Regarding this issue #1, for me personally this is a huge quality of life improvement, that is a relatively simple feature, and yet it allows to use Fortran modules more like Python. How this stacks up against some other features that we want to push, that's for a discussion to be had at #122.

**Comment by jme52** [comment]
1. I think this is a good idea.
    I worked for many years in a code where policy was that module procedures needed to start with the name or the acronym of the module.  This would make that code neater.

2. I like the `use, namespace :: utils` syntax.  I think `namespace` should be optional and disabled by default for backwards compatibility (sorry if I am stating the obvious).

3. I agree that `%` is the fortranic syntax.  However, its use may increase the complexity of the rules about scope of identifiers, which will need to be significantly rewritten. 

4. Would it be possible to simultaneously add the possibility of renaming modules (namespaces) as part of the use declaration?
    ```fortran
    use, non_intrinsic, namespace :: eh => electronic_history
    implicit none
    call eh%initialise_method
    ```

**Comment by everythingfunctional** [comment]
I like the idea of namespaces, and I like @jme52 syntax example, `use, namespace :: sm => some_module`, but one thing that should be pointed out is that you would not then be able to do `use, namespace :: sm => some_module, only: thing1, thing2`. Or at least I don't think you should be able to. Just thought this should be made explicit.

@jme52 , Why do you think using `%` increases the complexity of rules about scope? Do you have an example of that?

**Comment by certik** [comment]
@jme52, @everythingfunctional thanks for the feedback, I'll update the proposal.

Yes, `use, namespace :: eh => electronic_history` should be allowed. And `use, namespace :: sm => some_module, only: thing1, thing2` would not be allowed.

As part of this proposal, we also need to ensure we leave the door open if #86 is implemented, so that the syntax could be kept consistent.

**Comment by jme52** [comment]
@everythingfunctional, @certik: I don't know why renaming both the namespace and some of its names should not be allowed, would you mind to elaborate?

**Comment by everythingfunctional** [comment]
@jme52 , would that imply the following?
```
use, namespace :: sm => some_module, only: t1 => thing1
call sm%t1
call sm%thing2 ! this would not be allowed because thing2 wasn't brought into scope
call sm%thing1 ! this would not be allowed because thing1 was renamed
```

**Comment by certik** [comment]
The general idea that I have about this is the following equivalence between Python / Fortran:

| Python | Fortran |
| --- | --- |
| import A | use, namespace :: A |
| import A as B | use, namespace :: B => A |
| from A import x | use A, only: x |
| from A import x as y | use A, only: y => x |
| from A import * | use A |

If we allow `use, namespace :: sm => some_module, only: t1 => thing1`, then it should be a shortcut for:
```
! Fortran
use, namespace :: sm => some_module
use some_module, only: t1 => thing1

# Python
import some_module as sm
from some_module import thing1 as t1
```

Which answers @everythingfunctional's question. But I would not allow it, it seems confusing. One can always write the two statements as I did.

**Comment by everythingfunctional** [comment]
Given that
```
use, namespace :: sm => some_module, only: t1 => thing1
```
would be equivalent (if we allowed it) to
```
use, namespace :: sm => some_module
use some_module, only: t1 => thing1
```
then
```
call t1 ! this works
call sm%t1 ! this doesn't work because thing1 still has it's original name through the namespace
call sm%thing1 ! so this works
call sm%thing2 ! and this works
```

**Comment by certik** [comment]
@everythingfunctional yes, exactly.

**Comment by jme52** [comment]
@certik, @everythingfunctional: that's not what I was trying to propose.  To me,
```fortran
use, namespace :: sm => some_module, only: t1 => thing1
```
meant "from module `some_module`, which will be accessed through the `sm` name, import only `thing1`, which will be accessed through the name `t1`.  So only `sm%t1` would be available (nothing else `sm%<...>`, and nothing directly from `some_module`).

Forgetting about the `only`, do you envision being able to write something like this?
```fortran
use :: some_module
use, namespace :: sm1 => some_module
use, namespace :: sm2 => some_module
```
I am not strong on the theory of namespaces, so I don't know if this should be allowed, if it is a good idea, or something to be avoided.

**Comment by certik** [comment]
>  To me,
> 
> ```fortran
> use, namespace :: sm => some_module, only: t1 => thing1
> ```
> 
> meant "from module `some_module`, which will be accessed through the `sm` name, import only `thing1`, which will be accessed through the name `t1`. So only `sm%t1` would be available (nothing else `sm%<...>`, and nothing directly from `some_module`).

I would not do that. Just like you cannot import a derived type and rename or hide some of its parts. Python cannot do this either.

However, I agree that this syntax is confusing and meant a different thing to you than me. That's another reason not to allow it.

> Forgetting about the `only`, do you envision being able to write something like this?
> 
> ```fortran
> use :: some_module
> use, namespace :: sm1 => some_module
> use, namespace :: sm2 => some_module
> ```
> 
> I am not strong on the theory of namespaces, so I don't know if this should be allowed, if it is a good idea, or something to be avoided.

Yes, that should be perfectly fine. An equivalent Python code would be:
```python
from some_module import *
import some_module as sm1
import some_module as sm2
```

**Comment by jme52** [comment]
> > To me,
> > ```fortran
> > use, namespace :: sm => some_module, only: t1 => thing1
> > ```
> > 
> > 
> > meant "from module `some_module`, which will be accessed through the `sm` name, import only `thing1`, which will be accessed through the name `t1`. So only `sm%t1` would be available (nothing else `sm%<...>`, and nothing directly from `some_module`).
> 
> I would not do that. Just like you cannot import a derived type and rename or hide some of its parts.

The difference is that we can already rename in `use` statements via _rename_-_list_'s and _only_-_list_'s.  So if we want to allow for renaming namespaces (which you seem to support), I cannot see a reason why it should not be allowed.

> Python cannot do this either.

For any particular reason?

> However, I agree that this syntax is confusing and meant a different thing to you than me. That's another reason not to allow it.

I completely understand this meant something different to you, but I think that if any interpretation was included to the standard it would be crystal clear to users with a couple of examples.

To me it's rather the other way: the moment that fortraners who are already used to renaming variables see they can also rename modules they will try to do both at the same time, and they won't understand why they cannot do this.

> > Forgetting about the `only`, do you envision being able to write something like this?
> > ```fortran
> > use :: some_module
> > use, namespace :: sm1 => some_module
> > use, namespace :: sm2 => some_module
> > ```
> > I am not strong on the theory of namespaces, so I don't know if this should be allowed, if it is a good idea, or something to be avoided.
> 
> Yes, that should be perfectly fine. An equivalent Python code would be:
> 
> ```python
> from some_module import *
> import some_module as sm1
> import some_module as sm2
> ```
To me, multiple namespaces out of a single module sound more complicated: what happens to variables that are save'd inside of the module, does each of the namespaces get a copy of it, or do they share it?

A final comment: I am not sure that something being available or not in Python will be a strong argument for the committee.  I think it would be much more helpful to everybody an example could be presented where this is useful.

**Comment by certik** [comment]
> For any particular reason?

Yes --- I think the reason is that it fundamentally does not make sense. See below.

> To me it's rather the other way: the moment that fortraners who are already used to renaming variables see they can also rename modules they will try to do both at the same time, and they won't understand why they cannot do this.

There are two kinds of imports: one imports symbols from a module. We can do that already:

| Python | Fortran |
| --- | --- |
| from A import x | use A, only: x |
| from A import x as y | use A, only: y => x |
| from A import * | use A |

One can either import all symbols, or just some. And one can also rename symbols on import. I think there is no disagreement here. However, notice that neither Python nor Fortran allows you to import a derived type `x` and somehow rename its members. That fundamentally does not make sense. You can access the member `a` as `x%a`, or if you rename `x` at import time using `y%a`. But you cannot rename `a` to `b`. If you want to do that, you have to assign it as a variable:
```fortran
integer :: b
b = x%a
```

The other kind of import is to import the whole module as a namespace:

| Python | Fortran |
| --- | --- |
| import A | use, namespace :: A |
| import A as B | use, namespace :: B => A |

One can import it, and optionally rename it. You can view the module `A` as a derived type. Just like in the previous case it does not make sense to rename members of the derived type, it does not make sense to rename members of the module `A` if you are accessing them via `A%x`. If you want to use a different name, you have to do it using a variable, as in the previous case:
```fortran
type(x_t) :: y
y = A%x
```
Or you can use `associate`. 

Regarding using Python as an argument --- Python is just an example (but a good one!). Do you know of any language that would allow you to do what you are asking for? Also, do you have any use cases for it?

The use cases for my proposal above are many, e.g., https://github.com/fortran-lang/stdlib/pull/49, https://github.com/fortran-lang/stdlib/issues/99 as well as all the precedents in other languages. I think even you agree with my limited proposal based on your comments above. We are just catching up with other languages and we are not designing something new and untested.

The feature `use, namespace :: sm => some_module, only: t1 => thing1` on the other hand does not seem to have precedent in another language and it does not seem to be consistent with Fortran either, per my comment above. As such, that should give us a pause and we should step back and ask if it makes sense what we are designing. If you or others can motivate very well how this make sense and use cases, etc., I am not going to be blocking it. All I am saying is that this seems new and untested and seems inconsistent, so we better be sure we want that.

**Comment by certik** [comment]
The proposal was submitted as https://j3-fortran.org/doc/year/20/20-108.txt and it was discussed on Wednesday 2/26/2020 at the plenary meeting. I briefly described the feature in 2 minutes and we discussed for about 15 minutes. Here is the feedback from various members of the committee (thus not necessarily consistent):

* 
    - comment on SIN function: use GENERIC facility in current standard (can't you just make conflicting functions generic and rename on use?)
    - No big difference between module%foo and module_foo
    - this idea is not very compelling
    - do not like '%'; use ':' instead (do not like '.') possibly. Some compilers use colon in error messages and wouldn't use percent to indicate a name inside a module.
    - but colon probably can't be used either. backtick? apostrophe? Period won't work because use might not be private, and we might want nested modules in the future. 
    - prefers rename on use over namespacing anyways, even if it is there.



* this is great, fortran doesn't have great mechanisms for this.

* 
    - use Haskell language a lot which has this feature and it is useful
    - this idea should work for derived types and other entities also?
    - might be better as a "seller" of this idea to use entities other than procedures in use cases (derived type might be a better use case for this)

* 
    - While working with Modula: did something similar
    - but the main attraction of this idea is for compiler to track where an entity comes from

* 
    - not very compelling; renaming facility in current Fortran is ok
    - something like this might be more relevant in Python world where packages come from disparate places
    - stylistically nice, preferable if we have it, but hard to come up with a compelling use case. This is useful in python where there's lots of packages which might conflict, might be useful more as time goes on--we may have a future with packages & libraries. you can almost do everything now if you twist things to work, but it's not as convenient.


* 
    - this looks like a solution in search of a problem

* 
    - in relation to nesting of module entities, will direct qualification be possible? if you allow nesting there may be more than one unique way to get to the object you want.


* this promotes collaboration, and we have a hidden need for this. the language needs to evolve to facilitate collaboration.



It seems the main conclusion is to provide better use cases (with derived types, and other things) and to address the above objections in the proposal and resubmit next time.

**Comment by certik** [comment]
@zingale, @milancurcic I posted the Committee's feedback on this proposal above.

**Comment by ivan-pi** [comment]
Would wrapping all public subroutines and functions inside a module into a derived type be a simple way of achieving this "namespace" kind of functionality?

```fortran
module m_foo
    implicit none
    private

    public :: foo

    type :: t_foo
    contains
        procedure :: sqrt => sqrt_
        procedure :: cbrt => cbrt_
    end type

    type(t_foo), protected :: foo = t_foo()

contains

    function sqrt_(this,x) result(res)
        class(t_foo), intent(in) :: this
        real, intent(in) :: x
        real :: res
        res = sqrt(x)
    end function

    function cbrt_(this,x) result(res)
        class(t_foo), intent(in) :: this
        real, intent(in) :: x
        real :: res
        res = x**(1./3.)
    end function

end module

program test_foo
    use m_foo
    implicit none
    print *, foo%sqrt(4.), foo%cbrt(27.) !    2.00000000       3.00000000 
end program
```
Apart from functions under a generic interface or intrinsic functions, I think it would be doable with some smart preprocessing. There are some problems with this approach, e.g. the module and the type instance which wraps the module functionality must have different names (hence `m_foo`, `t_foo`, and `foo`); the type-bound procedures can in principle reuse intrinsic function names, but the function implementation requires a different name. 

The most annoying thing is that you have to duplicate all functions, if you want to have the possibility to call both specific functions or the type-bound procedures:
```
use m_foo, only: foo, cbrt  ! separate function needed for cbrt
```

---

The alternative route using abstract interfaces and procedure pointers doesn't seem to work, as the procedure pointer must be initialized with a variable that reduces to a constant expression.
```fortran
module m_foo
implicit none
abstract interface
  function sqrt_if(x) result(res)
    real, intent(in) :: x
    real :: res
  end function
end interface

procedure(sqrt_if), pointer :: psqrt => sqrt_ ! allowed

type :: t_foo
  procedure(sqrt_if), pointer, nopass :: sqrt => sqrt_ ! not allowed
end type
type(t_foo), protected :: foo = t_foo()

contains
  function sqrt_(x) result(res)
    real, intent(in) :: x
    real :: res
    res = sqrt(x)
  end function
end module
```
It is also not possible to use the structure constructor within a module as in `type(t_foo), protected :: foo = t_foo(sqrt=sqrt_)`. It works however if called in the main program as `type(t_foo) :: foo; foo = t_foo(sqrt=sqrt_)`.

**Comment by certik** [comment]
How would you do it for derived types, global arrays and other entities?

Besides that, your approach would work, but it is so complicated that it's better to not do it at all. For example this proposal allows to use a function both directly or via the module qualifier. In your approach I think you have to use different names.

**Comment by ivan-pi** [comment]
Essentially, you would need to place the entire contents of the module in the derived type and bind all the procedures. It is not a good solution, but for modules which only declare functions (e.g. blas or lapack) the functionality is similar (if you can live without being able to import the procedures directly).

I was hoping to show with my example, that similar functionality can already be achieved in Fortran by "misusing" derived types, but as you say the approach is so complicated, it would be better off to develop this into a true language feature.

**Comment by certik** [comment]
Yes, I agree. Thanks for bringing this idea up.

**Comment by wclodius2** [comment]
I have written a modified version of your 20-108.txt proposal. While I added some words that I think more clearly expressed the intent of the proposal, and added some use cases, the main change is I have replaced the NAMESPACE attribute for modules with a WITH attribute. I find WITH to be at least as intuitive as NAMESPACE and its brevity more user friendly FWIW I took the WITH attribute from Ada. Is it OK if I upload the revision as revised_namespace_proposal.txt?

**Comment by certik** [comment]
@wclodius2 yes, please do! Upload it as a new merge request. Let me know if you need any help with git or github. Thanks for bringing fresh ideas into this.

**Comment by wclodius2** [comment]
How do I do a merge request?

**Comment by milancurcic** [comment]
@wclodius2 Follow steps 2 and 3 from [here](https://github.com/j3-fortran/fortran_proposals/issues/22#issuecomment-657091589), but adjust the filename/directory for this proposal.

(merge request is same as pull request--the former is git/GitLab jargon, the latter is GitHub jargon :zany_face:)

**Comment by wclodius2** [comment]
I thought I did that properly, but I am having the following oddities:

1. My list of pull requests is showing one request with two files: proposals/default_optional_arguments/revised_proposal.txt, and proposals/namespace_modules/revised_namespace_proposal.txt,
2. The master directory for proposals/default_optional_arguments/ is not showing revised_proposal.txt though I thought Ondřej approved its pull request.
š. The list of pull request is showing not only my request, but an old one from February by Marshallward.

**Comment by milancurcic** [comment]
> 1. My list of pull requests is showing one request with two files: proposals/default_optional_arguments/revised_proposal.txt, and proposals/namespace_modules/revised_namespace_proposal.txt,

@wclodius2 I'm sorry. You did that properly, but my instructions were incorrect in the sense that they don't work if you have 2 concurrent PRs open. I only realize this now. To have 2 or more concurrent PRs, the commits (file additions) you made to your fork need to be made from separate branches. Currently, both proposals you committed to the master branch on your fork, and that's why both proposals now appear in #175. 

> 2. The master directory for proposals/default_optional_arguments/ is not showing revised_proposal.txt though I thought Ondřej approved its pull request.

This is expected. The PR has not been merged yet, so the file still exists only in your fork. The PR will stay open while we discuss it and edit it in the PR thread, and it will only be merged when it's ready to be uploaded to J3. @certik is this the correct workflow?

> 3. The list of pull request is showing not only my request, but an old one from February by Marshallward.

This is also expected. There can be many open PRs at any time.

So, how do you proceed? You need to commit the revised namespace proposal to a new branch on your fork. Perhaps the easiest way to do this is from the GitHub UI. Navigate [here](https://github.com/j3-fortran/fortran_proposals/tree/master/proposals/namespace_modules) and in the upper-right corner click on Add file -> Upload files. Because you don't have write permissions to this repo, GitHub will automatically create a new branch on your fork to which the files will be uploaded. Once done, navigate to https://github.com/j3-fortran/fortran_proposals and near the top there should be a yellow prompt from GitHub asking if you want to open a new PR from the new branch on your fork.

Give it a try, and if there are issues I'd be happy to get on a Zoom call with you to do it together via shared screen.

**Comment by wclodius2** [comment]
Don’t I have to first remove /revised_namespace_proposal.txt, from the initial PR?

**Comment by milancurcic** [comment]
@wclodius2 You don't have to do it first, but yes, `revised_namespace_proposal.txt` should be removed from the master branch in your fork (wcloudius2/fortran_proposals).

**Comment by wclodius2** [comment]
> On Jul 18, 2020, at 11:26 AM, Milan Curcic <notifications@github.com> wrote:
> 
> So, how do you proceed? You need to commit the revised namespace proposal to a new branch on your fork. Perhaps the easiest way to do this is from the GitHub UI. Navigate here <https://github.com/j3-fortran/fortran_proposals/tree/master/proposals/namespace_modules> and in the upper-right corner click on Add file -> Upload files.
> 
Did that and get the error message 
Uploads are disabled.
File uploads require push access to this repository.

I didn’t see anything indicating that GitHub automatically created a branch.
> Because you don't have write permissions to this repo, GitHub will automatically create a new branch on your fork to which the files will be uploaded. Once done, navigate to https://github.com/j3-fortran/fortran_proposals <https://github.com/j3-fortran/fortran_proposals> and near the top there should be a yellow prompt from GitHub asking if you want to open a new PR from the new branch on your fork.
> 
> 

Navigated to https://github.com/j3-fortran/fortran_proposals <https://github.com/j3-fortran/fortran_proposals> and didn’t see a yellow prompt.

Navigated to https://github.com/wclodius2/fortran_proposals/tree/master/proposals/namespace_modules <https://github.com/wclodius2/fortran_proposals/tree/master/proposals/namespace_modules> and there is no option to create a new branch and creating a new fork is disabled with the message “Cannot fork because you own this repository and are not a member of any organizations"

**Comment by milancurcic** [comment]
@wclodius2 Sorry, I thought that would work but it seems like GitHub UI is not as flexible for uploading new files as for editing files in-browser. I went ahead and uploaded your revision in #176 .

**Comment by wclodius2** [comment]
Thanks!

**Comment by wclodius2** [comment]
FWIW I have not seen any further response to this submission. I wonder whether that is because there has been no response or because @milancurcic made the submission and I cannot see responses to submissions by people other than myself? I noticed that some of the participants are labeled Collaborators or Contributors, but I lack such a label and wonder whether it affects what I see?

**Comment by septcolor** [comment]
Though I'm not sure if this has been discussed already, is it another approach to include the module name itself in the ONLY list (in a way similar to Julia)? For example, assuming that `mymod` has `integer :: foo, baa`, this statement

```
use mymod, only: mymod
```
allows only a qualified access to `foo` as
```
print *, mymod % foo   !! OK
print *, foo   !! error
```
At the same time, one can also import multiple entities as
```
use mymod, only: mymod, baa
use mymod, only: m => mymod, b => baa    !! rename
```

(The corresponding syntax in Julia is like `using Printf: Printf` and `using Printf: Printf, @printf`.)

**Comment by klausler** [comment]
I like septcolor's suggestion.  Would it be transitive -- i.e., if `mymod` has a `use hismod, only: hismod`, would `mymod%hismod%baz` work?

**Comment by certik** [comment]
I like @septcolor's suggestion also, I haven't though of that before.

@klausler yes, it would make sense to me for it to be transitive.

**Comment by septcolor** [comment]
@klauser I think the "transitive" behavior would be useful and intuitive. For comparison, I've tried this code in Julia, and it seems to behave as expected.
```
module mymod1
  foo = 100
end

module mymod2
  using ..mymod1: mymod1

  @show mymod1.foo   # OK: 100
  # @show foo        # Error: foo not defined
end

module mymod3
  using ..mymod2: mymod2

  @show mymod2.mymod1.foo   # OK: 100
  # @show mymod1.foo        # Error: mymod1 not defined
end
```

**Comment by wclodius2** [comment]
I am unenthusiastic about overloading the meaning of a keyword like ONLY, but do like the syntactic position of your suggestion. say
USE module-name, WITH: module-rename
strikes me as relatively intuitive

**Comment by septcolor** [comment]
(I am sorry if the following has already been discussed in other posts or thread...)

Another popular keyword used for this purpose is `import` (most notably Python), though this keyword is already used for different purposes (interface blocks) in Fortran. Because `import` is very common in other languages for importing modules, I guess another approach may be to "reuse" that keyword for importing a module name only (as a namespace), like `import mylib` at a top of a given scope (which I think corresponds to `use, with :: mylib` and `use, namespace :: mylib`).

**Comment by certik** [comment]
I like the word import the most. But as you said, it might be technically infeasible to use import for both. We need to see.

**Comment by gronki** [comment]
My opinion is that it is a bad idea to use word "import" for a completely
different meaning just because other languages do that. In Fortran as of
now including modules is accomplished by "use", not "import", and i think
it's important to stick to that to not introduce confusion.

Dominik

śr., 23 wrz 2020 o 02:44 Ondřej Čertík <notifications@github.com>
napisał(a):

**Comment by klausler** [comment]
`import` conflicts with existing usage; `with` doesn't make immediate sense to the new reader.

Consider something like `use :: mylib, ns => *` or `use, as(ns) :: mylib`.

**Comment by certik** [comment]
So far here are the options proposed so far, by category:

Add a keyword after `use` in line with the existing syntax for `intrinsic`:

* `use, namespace :: mylib`
* `use, import :: mylib`
* `use, with :: mylib`
* `use, as(ns) :: mylib`

Extend the `only` keyword:

* `use mylib, only`
* `use mylib, only:`
* `use mylib, only: mylib`

Extend the renaming `=>` syntax:

* `use :: mylib, ns => *`

Use two keywords:

* `use namespace mylib`

Use a new keyword entirely:

* `with mylib`

**Comment by certik** [comment]
I think we need to implement this in a compiler and actually use it.

Looking at these alternatives, the following starts to look nice:

* `use namespace mylib`

Easy to type, no `::`, no `,`, and this will be the most common use case. Maybe even like this:

* `use module mylib`

**Comment by klausler** [comment]
> I think we need to implement this in a compiler and actually use it.
> 
> Looking at these alternatives, the following starts to look nice:
> 
> * `use namespace mylib`
> 
> Easy to type, no `::`, no `,`, and this will be the most common use case. Maybe even like this:
> 
> * `use module mylib`

Doesn't work with fixed-form source, unfortunately.  `USENAMESPACEMYLIB` is ambiguous, and unlike `MODULEPROCEDUREFOO`, not clear from the context.

**Comment by certik** [comment]
> Doesn't work with fixed-form source, unfortunately. `USENAMESPACEMYLIB` is ambiguous, and unlike `MODULEPROCEDUREFOO`, not clear from the context.

I forgot about that. Another example for #108. One option would be to only introduce this syntax in free form, it would not be available in fixed form.

**Comment by klausler** [comment]
> > Doesn't work with fixed-form source, unfortunately. `USENAMESPACEMYLIB` is ambiguous, and unlike `MODULEPROCEDUREFOO`, not clear from the context.
> 
> I forgot about that. Another example for #108. One option would be to only introduce this syntax in free form, it would not be available in fixed form --- but that's a bigger decision to be made separately.

All Fortran semantics are orthogonal to the source form.  And it's easy to accommodate fixed form here -- you listed many workable alternatives above.

**Comment by wclodius2** [comment]
Before implementing it it needs to be specified well enough that you know what you are implementing. First in Fortran punctuation is very useful, particularly to remove ambiguities in fixed source form. Second you will want to accommodate the module-nature keywords: intrinsic and the rarely used nonintrinsic. You will also probably want to accommodate module renaming. I am inclined to a single new keyword as it can be unambiguous and reduces typing

* namespace, intrinsic:: iso_fortran_env => iso
* with, intrinsic:: iso_fortran_env => iso
* using, intrinsic:: iso_fortran_env => iso

**Comment by milancurcic** [comment]
Here's one more, inspired by @klausler but consistent with `use ..., only:`:

```
use iso_fortran_env, as: env
```

**Comment by ivan-pi** [comment]
Stavros (can't find his github name) left a comment in the monthly call, how would the namespace syntax work together with submodules?

**Comment by everythingfunctional** [comment]
> Here's one more, inspired by @klausler but consistent with `use ..., only:`:
> 
> ```
> use iso_fortran_env, as: env
> ```

Stylistically I prefer `use iso_fortran_env, as: env` over `use, namespace :: env => iso_fortran_env`, but the later is more consistent with existing syntax, and simplifies in a more consistent and concise way.

**Comment by everythingfunctional** [comment]
> Stavros (can't find his github name) left a comment in the monthly call, how would the namespace syntax work together with submodules?

Would it be relevant? Currently, consumers of a module can't tell if it uses submodules, so there's no issue on that side. And anything in scope in a module is implicitly in scope in a submodule, so I don't think there's an issue on that side either.

**Comment by milancurcic** [comment]
Tagging @smeskos here.

**Comment by certik** [comment]
I like `use iso_fortran_env, as: env`, I think we all do.

The question is how to simplify:
```
use iso_fortran_env, as: iso_fortran_env
```
Some options:
* `use iso_fortran_env, as`
* `use iso_fortran_env, as:`
* `use iso_fortran_env, as: *`
* `use iso_fortran_env, as is`
* `use iso_fortran_env, as module`
* `use iso_fortran_env, as namespace`

**Comment by septcolor** [comment]
Just FWIW, Chapel uses an empty identifier list after `only` (https://chapel-lang.org/docs/primers/modules.html) to import only the module name, which may translate to

```
  use mymod, only:   !! nothing written here
```
(which is equal to `import mymod;` that allows only a qualified access of module entities).

Nim uses a similar approach by excluding all entities in the module via the keyword `nil` (e.g., `from mymod import nil`), which may correspond to

```
  use mymod, only: none
```

(but this approach needs an additional keyword like "none" etc...)

Julia uses both `using` and `import`, where `using` imports all the exported module entities by default, while `import` imports only the module name.

**Comment by certik** [comment]
@septcolor you are right. Here is a comparison table:

| Python | Julia | Chapel | Fortran |
| --- | --- | --- | --- |
| import A | import A | use A only | use A, only: |
| import A as B | import A as B | use A as B only |  use B => A, only: |
| from A import x | using A: x | use A only x | use A, only: x |
| from A import x as y | import A: x as y | use A only x as y | use A, only: y => x |
| from A import * | using A | use A | use A |

Note 1: Julia's syntax for `import A as B` and `import A: x as y` is only implemented at a PR (https://github.com/JuliaLang/julia/pull/37396), but not merged yet.

Note 2: One can use `import A: x` instead of `using A: x` in Julia.

Note 3: Chapel's syntax `use A as B only` isn't mentioned in the docs, but it seems that would be the natural way (I haven't tested it).

Note 4: Chapel also has an `import` syntax, but it seems redundant (?) to the `use` syntax above.


Given this and the Chapel's precedent, I now prefer:

* `use A, only:`
* `use B => A, only:`

**Comment by klausler** [comment]
`use A, only: ! empty` to get a namespace is not at all clear, and it doesn't allow for the otherwise orthogonal creation of both a namespace as well as unqualified importation of some names.

**Comment by milancurcic** [comment]
I don't like `use A, only: ` because to a naive reader, it's not clear what it means. To me (experienced reader), it's still awkward and seems like it's trying to fit an existing syntax that is used for something else.

Of @certik's alternative syntaxes, I like this one best:

```
use iso_fortran_env, as is
```

**Comment by wclodius2** [comment]
FWIW having a keyword after `only:` is prone to namespace conflicts. For example, my `stdlib_bitsets` module defines a function `none`.

**Comment by septcolor** [comment]
@klausler As far as I experimented, it seems possible to import the module name as well as module entities by writing two statements, e.g. (in Chapel's case):

```
use Time only;
use Time only Timer as Foo;

proc test()
{
    var timer = new Time.Timer();
    writeln( timer.type:string );  // Timer

    var foo = new Foo();
    writeln( foo.type:string );  // Timer
}

test();
```
but I agree that `use mymod, only:  !! nothing here` looks somewhat awkward (at least initially). If there were some keyword representing "nothing", we could use it here, but @wclodius2 says, `none` etc can conflict with existing symbols.

Personally, I prefer using the keyword `import` most (which is very natural in its meaning), but it is "reserved" by the use in interface blocks. I think one could define "except this, except that" rules to enable overloading the meaning of `import`, but I'm afraid this would introduce additional complications (possibly both for compiler implementation and user experiences)...

**Comment by klausler** [comment]
For what it's worth, Haskell has `import [qualified] ModuleName [as Alias] [[hiding] (names)]`.

Qualified names are available with and without `qualified` -- the module name or its alias always becomes available as what you all are calling a "namespace".  When `qualified` doesn't appear, the imported names are also available without qualification.

The list of imported names can be absent, in which case all exported names from the module are imported.  And it can be negated with `hiding` to import everything *but* some names.

Haskell also allows multiple modules to be imported into the same alias, so long as there are no conflicts.  And module aliases can be *exported*, too, for use in other modules.  This is useful -- it lets one compose a new module using names imported from others.

**Comment by smeskos** [comment]
@everythingfunctional:  
> Would it be relevant?

Perhaps not. I am trying to figure this out. Please see the two examples below. Do they make any sense or my confusion is irrelevant to the namespace proposal?

First example with submodules;
We have the following module/submodule system:
                                                                         module Base
submodule A,                                                                                           submodule B
submodule A1,   submodule A2

Declaration: 
submodule (Base)A,            submodule(Base)B
submodule(Base : A)A1,   
submodule(Base : A)A2

```fortran
program main
use, namespace :: Base
!will Base be able to see every submodule?
call Base%A%A1%mySubroutine(args, ...)
end program
```

Second example:

module All, module A, module B, module C, module D, etc...
```fortran
module All
use, namespace :: A, B, C, D,...
end module All
! Is the following correct?
program main
use, namespace ::all
call all%A%mySubroutineA()
end program main
```

**Comment by everythingfunctional** [comment]
Submodules aren't supposed to be publicly knowable, so if `mySubroutine` is publicly available from `Base`, then it would be accessible as `Base%mySubroutine` and there would be no need for any reference to the submodule. If something is not declared in the module, it is not publicly accessible.

**Comment by wclodius2** [comment]
As is common with arguments about syntax, we are "bike shedding". Everyone understands syntax, so everyone has a different opinion on it. I can come up with a large number of syntaxes. For example a standalone keyword:

    namespace, intrinsic:: env=>iso_fortran_env
    with, intrinsic::  env=>iso_fortran_env
    using, intrinsic:: env=>iso_fortran_env
    import, intrinsic:: env=>iso_fortran_env
    employ, intrinsic:: env=>iso_fortran_env
    source, intrinsic:: env=>iso_fortran_env
    qualified, intrinsic:: env=>iso_fortran_env
    as is, intrinsic:: env=>iso_fortran_env

many of the same keywords could be used as attributes in the early part of a use statement, but then the standard has to explain why there is no rename or only list;

    use, namespace, intrinsic:: env=>iso_fortran_env
    use, with, intrinsic :: env=>iso_fortran_env
    use, source, intrinsic :: env=>iso_fortran_env
    use, qualified, intrinsic :: env=>iso_fortran_env
    use, as is, intrinsic :: env=>iso_fortran_env

or they could be used as replacements for `only`

    use, intrinsic:: iso_fortran_env, namespace: env
    use, intrinsic:: iso_fortran_env, with: env
    use, intrinsic:: iso_fortran_env, as is: env
    use, intrinsic:: iso_fortran_env, qualified: env

I am sure that with more time I could come up with more examples.

**Comment by milancurcic** [comment]
@wclodius2 I don't think we're bike shedding. Syntax and choice of keywords are not irrelevant details. They have a direct impact on the programmer, what they're writing, and how they feel about it. I think that the new syntax should:

* Be as minimal as possible
* Be as intuitive as it can be to a naive reader
* Be as consistent with existing syntax as possible
* Spark joy

The optimal solution is not easy to come up with and it's impossible to make everybody happy. All these tedious discussions about words and punctuation and what's intuitive or not are necessary and important part of the work. The more people that provide feedback the more clear we will get about the top candidates.

**Comment by certik** [comment]
> it doesn't allow for the otherwise orthogonal creation of both a namespace as well as unqualified importation of some names.

Neither Python, Julia or Chapel allow that as far as I know with a single statements. @septcolor gave an example in Chapel, in Python you have to do:
```
import numpy
from numpy import sin
```
So in Fortran it would become:
```
use numpy, only:
use numpy, only: sin
```

I agree with @klausler that it looks a bit confusing. Even without the `:` it looks weird to me: `use numpy, only`. The Chapel version, on the other hand, looks good: `use numpy only` and `use numpy only x`, but we can't use this without a comma due to ambiguities in the fixed form.

@wclodius2 is right that it might seem like we are "bike shedding", but in order to move this proposal further, the syntax is important. @milancurcic just replied above in similar spirit.

We agreed at the call that we will do a survey at the Fortran Discourse and ask people for preferences.

**Comment by wclodius2** [comment]
There are two orthogonal choices in the syntax: first what word to use as the keyword, and second where to put the keyword. So far I have seen the following suggestions for keywords:

`namespace` - main problem is its length, but users of C++, where `namespace` behaves like Fortran's `use`, may find this usage counterintuitive
`with` - main advantage is its length, some find it unintuitive
`using` - similar to use, slightly suggests requiring the module name to access
`import` - used for this in Python, but already a keyword with different semantics
`employ` - a synonym to use
`source` - slightly suggestive of the semantics
`qualified` - one of its meanings is suggestive of the semantics, but not the most commonly used meaning, as long as namespace
`as` with `as is` - one for renaming the other for non-renaming. What happens if you want to rename to is?
`ns` - an abbreviation for namespace that is similar to nanosecond
`only` - only as a use statement terminator, overloads its current meaning

Are there any names I missed?

I know of three possible places to put the keyword

At the beginning, making its own statement, i.e,
keyword[[, module-nature]::] [local-module-name=>]module-name
this is what most of the languages cited do

After the use keyword
use [[, keyword][, module-nature]::] [local-module-name=>]module-name
or
use [[, module-nature][, keyword]::] [local-module-name=>]module-name
It will be tempting to lump the keyword with intrinsic and nonintrinsic in module-nature, in which case the standard needs to explain that there is no only or rename list.

Replacing the only keyword
use[[, module-nature]::] module-name, keyword[:local-module-name]
making a third form of a `use` statement

**Comment by smeskos** [comment]
@wclodius2 
> Are there any names I missed?

The word _name_ itself as a verb or the word _rename_.  

However, I find all of the proposed keywords as non-Fortranic way of doing things. With the only exception perhaps of the original proposal of _namespace_. 
An example of my personal favorite style of implementing this:
```fortran
use mathematics, only: cdif=>central_difference
call cdif()
!to
use, namespace, math=>mathematics, only:cdif=>central_difference
call math%cdif()
```
The keyword **namespace** will be employed to distinguish between using or not using the name of the module and the symbol `=>` optionally if we want to rename locally the name of the module.

Another alternative to **namespace** could be **modulename**, only because the former may cause confusion with the C++ keyword, and why not to reserve it for a possible future similar implementation.
Therefore, the alternative to my previous example:
```fortran
use, modulename, math=>mathematics, only:cdif=>central_difference
call math%cdif()
```

**Comment by FortranFan** [comment]
I will request everyone to take a look at my [**comment** ](https://github.com/j3-fortran/fortran_proposals/issues/86#issuecomment-554205627)in another thread #86 that I think is related to this one.  As I suggest therein, I personally would preface NAMESPACE to reflect a **formal** concept to be introduced in the Fortran standard at some later stage in the advancement of this language that always holds tremendous promise as a tool for scientists and engineers but which is perpetually trying to "catch up" to the state-of-the-art in technical computing.  Meaning the language and its bearers haven't fully caught up yet to the concept of namespace and its benefits though Fortran took a big step toward it with MODULEs back in 1991 with Fortran 90 publication.

In the context of this particular proposal in this thread, I think the search is for a way to **decorate** module entities suitably during their "consumption" via USE association.

And I think the _Fortrannic way_ will be to introduce both an ATTRIBUTE and a STATEMENT for the same.  And where the options gain a FUNCTION-like syntax with optional arguments within parenthesis.

There are several examples of this in the Fortran standard: ALLOCATABLE, BIND, etc.  `BIND` seems to me to be quite apt here for understanding a facility that is both an ATTRIBUTE and a STATEMENT.  Note the token `BIND` in my simple-minded thinking can be viewed as a verb or a command or an instruction which is what seems to be the need here.

So thinking along the same lines, an alternate suggestion will be to use `DECORATE` as the token:

```Fortran
use, decorate :: utils 
..
call utilsXsavetxt(..)
```
where X is the separator yet to be decided; it can be `%` (as in `call utils%savetxt(..)`) or it might be double-colon (as in `call utils::savetxt(..)`), etc.

And additional options with (re)naming can be introduced as
```Fortran
use, decorate( name="ut" ) :: utils 
..
call utXsavetxt(..)
```

With the STATEMENT form, the expected syntax is
```Fortran
use utils
..
decorate :: utils
..
call utilsXsavetxt(..)
```

Applying the same to @smeskos 's examples above, I would expect the syntax to be

```Fortran
use, decorate(name="math") :: mathematics, only : cdif => central_difference
call mathXcdif()
```

And if the optional `name` is not employed,
```Fortran
use, decorate :: mathematics, only : cdif => central_difference
call mathematicsXcdif()
```

With the example in the J3 paper (https://j3-fortran.org/doc/year/19/19-246.txt), the syntax can look like so:
```Fortran
    use, decorate :: math
    use, decorate(name="np") :: numpy
    use, decorate( name="sym") :: sympy
    ...
    e1 = npXsin(npXpi)      ! NumPy expression
    e2 = mathXsin(mathXpi)  ! Built-in Python math expression
    e3 = symXsin(symXpi)    ! SymPy expression
```

and
```Fortran
    use, decorate :: lapack
    ..
    call lapackXdgeev('N', 'V', n, At, lda, wr, wi, vl, ldvl, vr, ldvr, &
     work, lwork, info)
    ..
    call lapackXdgetrf(n, n, Amt, lda, ipiv, info)
    ..
```

**Comment by gronki** [comment]
I find "decorate" concept elegant, except I am not sure why the alternative
name is to be put in quotes. In free form, spaces or special characters are
not allowed in names afaik, while string literal implies that any
characters should be valid. I think that

use, decorate(name = lpk) :: lapack

would be more logical.

When I look at this syntax, it really looks like Fortran to me. Good shot
FortranFan.
Also because one USE is needed to include one module, while => operator
usually appears in constructs that can be used in sequences separated by a
comma. So

use, namespace :: np => numpy

naturally encourages user to do

use, namespace :: np => numpy, lpk => lapack

which would be too much of a stretch considering current USE syntax.

I also would agree that using % (so the same operator used to access
derived type components) should be very carefully discussed. I would also
prefer to have a distinguishment between accessing module and derived type
components. :: seems to be another good candidate. While I love Python and
in that language no distinguishment makes sense (its an interpreted
language), I would say in Fortran I would rather know whether it's a
derived type or just a namespace. There is enough confusion as is with
parethesis () used both for functions and arrays. Let's not repeat that
mistake.

Dominik

sob., 26 wrz 2020 o 19:30 FortranFan <notifications@github.com> napisał(a):

**Comment by milancurcic** [comment]
I also like @FortranFan's idea of specifying the namespace name in parentheses following the keyword. Also with @gronki that the new name shouldn't be a string literal. 

I don't like the word "decorate" here. To me it means to add functionality to something which I don't think is what we're doing.

**Comment by FortranFan** [comment]
@gronki, @milancurcic ,

I'm ok with the synonym to be not a string literal i.e., a *`default-char-constant-expr`* per the standard.  I had suggested as much based on what I thought I had heard previously with certain compiler implementations and how they parse the FUNCTION-like approach with the use of parenthesis in ATTRIBUTEs and STATEMENTs.   This is where an implementation or two, as suggested by @certik, will come in handy: it can help work through such options.

@milancurcic ,

Re: "I don't like the word "decorate" here. To me it means to add functionality to something which I don't think is what we're doing.":

* note this proposal is indeed adding a functionality, albeit it is in the form of an adornment or an embellishment.  What might previously be a use-associated entity `FOO` from `MY_MOD` module will no longer have its first-class (or an independent) identity that it has with current standard.  Meaning, there can no longer be a reference to `FOO` in that scope.  It must be "adorned" as something like `MY_MOD::FOO` (or `MY_MOD%FOO`, depending on the separator chosen.)

Then, in the dictionary sense of the term, this is nothing but decoration.  That's why I suggested that token.  It seems to better capture the essence of the functionality being proposed here than all the alternatives suggested upthread.

But if it's too verbose, another alternative is **`ADORN`**.  But if people starting getting into further nits with the choice of the word, that will confirm @wclodius2 's point earlier about the bikeshedding here.

**Comment by gronki** [comment]
Yeah perharps the choice of a word is a bit of a detail. From my
perspective as a non-native English speaker, the word should be
commonly used in international English (never seen "adorn" before) and
not to be confused with something else (like IMPORT proposed in
another context). I am sure there is one than more word that fits this
criteria :)

Dominik

niedz., 27 wrz 2020 o 19:02 FortranFan <notifications@github.com> napisał(a):

**Comment by milancurcic** [comment]
@FortranFan Thank you, only after your detailed explanation I was able to understand--you're not decorating the module itself, but the public entities in the module. Now it makes more sense.

I'd still advise against it because of how easily it can be confused with decorating as a software design pattern. For example, if I was teaching this concept, I'd have to caveat it with something like "If you've used decorators in other languages before, be careful--this does something completely different", and then refer to it as "namespace" in the rest of the material. :)

This is also a good example of why I think the choice of the word is important.

**Comment by septcolor** [comment]
Dear @certik,

> > it doesn't allow for the otherwise orthogonal creation of both a namespace as well as unqualified importation of some names.
> 
> Neither Python, Julia or Chapel allow that as far as I know with a single statement. 

Julia seems to allow this with one line, e.g. to import "LinearAlgebra" (a module or package name) and "svd" (a function name) only as

```
using LinearAlgebra: LinearAlgebra, svd 
```

> Note 1: Julia's syntax for import A as B and import A: x as y is only implemented as a PR (JuliaLang/julia#37396) from 3 weeks ago (!), but not merged yet.

Thanks for the info! As for Note 4 (why Chapel introduced "import" in addition to "use"), I came across long discussion pages before, but was not able to catch the core reason very well ... (it might be related to automatic search for packages and modules).

> (FortranFan) X is the separator yet to be decided; it can be % (as in call utils%savetxt(..)) or it might be double-colon (as in call utils::savetxt(..)), etc.

`::` looks also nice :)

---------------------

And one more possibility may be this form (taken from the Fortran Wiki)? (sorry if this is already discussed...)
http://fortranwiki.org/fortran/show/Extension+to+USE+statement

```
USE [[, module-nature] ::] module-name [AS use-name] [, ONLY:only-list]

Without “AS use-name” it is a regular USE statement. However, when the “AS use-name” is used, then the entities defined in module “module-name” are accessed only with “use-name%entity-name” syntax.

Eg: USE hitechlib AS ht
```

**Comment by FortranFan** [comment]
@septcolor wrote: 

> ..
> And one more possibility may be this form? ..
> .. USE hitechlib AS ht

As mentioned upthread, punctuation or a separator between conjunctions is needed to support the 2 sources forms in Fortran, and that's why this does not work.

**Comment by septcolor** [comment]
@FortranFan 
Thanks, then maybe we can do: `USE hitechlib, AS: ht` (?)

As another point to be considered: I am wondering if modules import different modules in a cascaded(?) way, do various proposed syntaxes work?

**Comment by certik** [comment]
Thanks everybody for your input and making progress on this issue.

@septcolor raises an important thing to consider: #86 (nested modules). As posted there, this might work:
```
use, namespace :: mylib%a
...
call mylib%a%something()
```

I like the `::` option instead of `%` also, but I don't know if it will work syntax wise for nested modules:

```
use, namespace :: mylib::a
...
call mylib::a::something()
```

Don't know if there could be issues parsing it.

**Comment by wclodius2** [comment]
There seems to have been a consensus that there should be a poll/vote on the preferred syntax. As there has been a significant pause in postings on this issue, I think it is an appropriate to summarize our current syntactic options to prompt more suggestions for options. There are four orthogonal choices in the syntax: first what word to use as the keyword, second where to put the keyword, third, how to do module renaming, and fourth, what to use as the accessor "operator". So far I have seen the following suggestions for keywords:

adorn - a shorter version of decorate
as or as is - one for renaming the other for non-renaming. What happens if you want to rename to is?
decorate - slightly long, has a dirfferent meaning in software engineering
employ - a synonym to use
import - used for this in Python, but already a keyword with different semantics
namespace - main problem is its length, but users of C++, where namespace behaves like Fortran's use, may find this usage counterintuitive
ns - an abbreviation for namespace that is similar to nanosecond
only - only as a use statement terminator, overloads its current meaning
qualified - one of its meanings is suggestive of the semantics, but not the most commonly used meaning, as long as namespace 
source - slightly suggestive of the semantics
using - similar to use, slightly suggests requiring the module name to access
with - main advantage is its length, some find it unintuitive

Are there any names I missed? With this many options it might be best to do a ranked choice vote.


I know of three possible places to put the keyword

At the beginning, making its own statement, i.e,
keyword[[, module-nature]::] ...
this is what most of the languages cited do

After the use keyword
use [[, keyword][, module-nature]::] ...
or
use [[, module-nature][, keyword]::] ...
It will be tempting to lump the keyword with intrinsic and nonintrinsic in module-nature, in which case the standard needs to explain that there is no only or rename list.

Replacing the only keyword
use[[, module-nature]::] module-name, keyword[:local-module-name]
making a third form of a use statement


I have noted three ways to do module renaming:

First use a syntax based on the rename list of the use statement [local-module-name=>]module-name, e.g.,
keyword[[, module-nature]::] [local-module-name=>]module-name
use [[, keyword][, module-nature]::] [local-module-name=>]module-name
use [[,module-nature]::][local-module-name=>]module-name, keyword

Second after the colon after the module name where the keyword replaces only
use [[,module-nature]::]module-name, keyword[: local-module-name]

Third in a parenthesized rename after the keyword
keyword[([name=]local-module-name)][[, module-nature]::] module-name
use [[, keyword[([name=]local-module-name)]][, module-nature]::] module-name
use [[,module-nature]::]module-name, keyword[([name=]local-module-name)]


I have noted two suggestions for the accessor operator
%, used as the accessor for derived types
::, used extensively in the specification-part, but not in the execution-part of Fortran programs/subprograms. The main question is would its use cause what is intended to be an executable statement, to be interpreted as a specification statement, i.e., could a module name be the same as the start of a specification statement.
In principle other characters could also be used: `, ~, @, ^, |, and \ are ASCII characters not used by Fortran, and, I believe, not used by the common pre-processors.


FWIW I tend to prefer the shorter keywords, but dislike overloading the meaning of "only" and "import". so my first preference is "with" followed by "using" and "adorn". I believe modifying the use statement to incorporate the keyword, will more complicate the standard than having a keyword statement, so I prefer a keyword statement. Because I don't want a modified use statement, I don't want the rename after the colon after the keyword. I don't have strong feelings about keyword[([name=]local-module-name)] versus [local-module-name=>]module-name. I prefer the accessor to be the same as for derived types, i.e., "%", but "::" would also be acceptable.

**Comment by milancurcic** [comment]
Thank you for the summary @wclodius2, it's very helpful. Here's my current preference:

1. **What word to use**: `as` (renaming) and `as is` (non-renaming)
2. **Where to put the keyword**: Option 3: `use[[, module-nature]::] module-name, as[: local-module-name]`
3. **How to rename modules**: `use[[, module-nature]::] module-name, as is`
4. **What to use as accessor operator**: Aesthetically I like `::` best. If it's problematic due to issues you raise, I like either `%` or `@`. Others I don't like.

Examples:

* `use stdlib_stats, as: stats`
* `use stdlib_stats, as is`

Addendum to the syntax proposal which would allow orthogonal imports as @klausler suggested: It's possible to use the `only` keyword following the `as: local-module-name`, so the syntax becomes:

```
use[[, module-nature]::] module-name, as[: local-module-name][, only: [local-entity-name =>] entity-name]
```

Example:

```
use stdlib_stats, as: stats, only: mean, var
```

This statement would import `stdlib_stats` as a `stats` namespace (with all public entities therein available via the namespace), and also import `mean` and `var` into the global namespace.

To prevent writing confusing code, `as` would have to appear before `only` if both are used. So

```
use stdlib_stats, only: mean, var, as: stats
```

shouldn't be allowed IMO.

**Comment by klausler** [comment]
I don't see why syntax is needed to enable a namespace without renaming.  That could be automatic in the absence of a namespace name.  None of gfortran, ifort, nagfor, or xlf allow a module name to be used in a scope for anything else after it appears on a `USE` statement.

**Comment by milancurcic** [comment]
I didn't think about that. It does seem like the most elegant solution. Simply make it available as a namespace if you don't need it renamed. It's backward compatible because, as you say, it has been impossible to reference a module name after a `use` statement. I like it.

**Comment by wclodius2** [comment]
First Fortran module names tend to be long to reduce the possibility of name clashes so most of the time users would find renaming useful though it could be

use[(name=local-module-name)][[,module-nature]::]module-name..;.

Second some users will want the processor to enforce the use of the module name as a qualifier which in turn requires the use of a different syntax to let the processor know that that is a requirement.

**Comment by klausler** [comment]
If `::` is the separator, then `INTEGER::X=666` would be ambiguous if `INTEGER` were a namespace.

**Comment by milancurcic** [comment]
> some users will want the processor to enforce the use of the module name as a qualifier which in turn requires the use of a different syntax to let the processor know that that is a requirement.

Indeed, once you can namespace, simple and common names will proliferate in modules and the user should have a way to not have them pollute the global namespace.

**Comment by klausler** [comment]
My preference is to replace the current

```
R1409 use-stmt is USE [ [ , module-nature ] :: ] module-name [ , rename-list ]
               or USE [ [ , module-nature ] :: ] module-name , ONLY : [ only-list ]
```

with

```
R1409 use-stmt is USE [ [ , module-nature ] :: ] [namespace-name=>] module-name [ , rename-list ]
               or USE [ [ , module-nature ] :: ] [namespace-name=>] module-name , ONLY : [ only-list ]
```

with automatic creation of a namespace when an explicit `[namespace-name=>]` is missing.

In some far-off future when `module-name` becomes `module-name [( module-actual-arg-list )]`, automatic creation of a namespace wouldn't happen when actual arguments appear, of course.

**Comment by certik** [comment]
@klausler That seems to be a very similar idea to what I proposed above https://github.com/j3-fortran/fortran_proposals/issues/1#issuecomment-699120529, but I thought you didn't like it. Are you proposing exactly the table in that comment, or are you proposing the following table (in other words, do you want to allow `use A, only:` or not):

| Case | Python | Julia | Chapel | Fortran |
| --- | --- | --- | --- | --- |
| 1. | import A | import A | use A only | use A |
| 2. | import A as B | import A as B | use A as B only |  use B => A |
| 3. | from A import x | using A: x | use A only x | use A, only: x |
| 4. | from A import x as y | import A: x as y | use A only x as y | use A, only: y => x |
| 5. | from A import * | using A | use A | use A |

Furthermore, in all cases 1.-5., the module name `A` (or `B` in 2.) will always be created as a namespace, in addition to importing all the symbols from it in 1., 2. and 5. (and only `x` or `y` in 3. and 4.).

**Comment by klausler** [comment]
Yes, the syntax already permits `use :: m, only:`.  But I don't see a need for a new `, only: *`, and I don't think it's a good idea to force the use of new syntax to avoid a warning on existing code that shouldn't need to be changed.

**Comment by certik** [comment]
Ok, so to summarize, you are proposing this:

| Case | Python | Julia | Chapel | Fortran |
| --- | --- | --- | --- | --- |
| 1. | import A | import A | use A only | use A, only: |
| 2. | import A as B | import A as B | use A as B only |  use B => A, only: |
| 3. | from A import x | using A: x | use A only x | use A, only: x |
| 4. | from A import x as y | import A: x as y | use A only x as y | use A, only: y => x |
| 5. | from A import * | using A | use A | use A |

Where the cases 1. and 2. are exactly equivalent to Python, but cases 3., 4. and 5. are a "superset" of Python: they import the `x`, `y` or all symbols as in Python respectively, but also in addition make the module name `A` available as a namespace in Fortran (which does not happen in Python).

I like this proposal.

**Comment by klausler** [comment]
> Ok, so to summarize, you are proposing this:
> 
> Case	Python	Julia	Chapel	Fortran
> 1.	import A	import A	use A only	use A, only:
> 2.	import A as B	import A as B	use A as B only	use B => A, only:
> 3.	from A import x	using A: x	use A only x	use A, only: x
> 4.	from A import x as y	import A: x as y	use A only x as y	use A, only: y => x
> 5.	from A import *	using A	use A	use A
> Where the cases 1. and 2. are exactly equivalent to Python, but cases 3., 4. and 5. are a "superset" of Python: they import the `x`, `y` or all symbols as in Python respectively, but also in addition make the module name `A` available as a namespace in Fortran (which does not happen in Python).
> 
> I like this proposal.

I don't know any of those languages well enough to know whether there's a correspondence or not.

**Comment by klausler** [comment]
The harder part of this proposal remains to be discussed, too.  Of all of the places in the Fortran grammar where "name" or "*-name" or "name-list" (&c.) appears, which will be replaced with "qualified-name"?  Or will all names be potentially qualifiable by default, with some possible exceptions that will have to become "unqualified-name"?  There's a *lot* of ways that names can be used in the language and the due diligence here is going to be a lot of work.

And some cases are not obvious.  Will `MODULE PROCEDURE NS%FOO` be allowed as a top-level program unit?  Is `EXTERNAL :: NS%FOO` ok?  Would `VOLATILE NS%FOO` make that variable volatile in the local scope?  Will lexers have to accept `NS%KP_"string"`?  (I'm not looking for answers, just listing some cases where a decision to allow a qualified name or not might not be clear for syntactic reasons or semantic implications.)  Somebody's going to have to sign up to work this out in order to have a complete proposal for a language feature.  This is an area where implementation first would drive resolution, too.

**Comment by certik** [comment]
Thanks Peter. My goal is to get it pre-approved by the committee, saying "yes, we want this feature, as long as all the issues can be satisfactorily resolved". The minute it does, I am sure multiple people from the community will help. Obviously nobody wants to put in weeks of work if the committee decides they don't want this feature in the first place.

I plan to prototype this in a compiler.

**Comment by FortranFan** [comment]
@klausler  wrote Oct. 1, 2020 6:48 PM EDT:

> The harder part of this proposal remains to be discussed, too. ..

I agree. 

As I alluded to earlier, Fortran language can do better with gaining the formal concept of namespaces.  Ideally, features such as this are best handled as part of that effort.

Considering the complexities and constraints involved with piecemeal attempts, a prototype implementation for the formal namespace concept itself will be nice.  The aspects such as this thread can be subparts of such an initiative.

**Comment by klausler** [comment]
> @klausler wrote Oct. 1, 2020 6:48 PM EDT:
> 
> > The harder part of this proposal remains to be discussed, too. ..
> 
> I agree.
> 
> As I alluded to earlier, Fortran language can do better by gaining the formal concept of namespaces. Ideally, features such as this are best handled as part of that effort.
> 
> Considering the complexities and constraints involved with piecemeal attempts, a prototype implementation for the formal namespace concept itself will be nice. The aspects such as this thread can be subparts of such an initiative.

How do your namespaces differ from modules?

**Comment by septcolor** [comment]
Because I have no experience of writing a compiler, I strongly feel that "noob" opinions and proposals (including me) is actually very dangerous (because I cannot imagine such issues very soon). At the very least, I think people with sufficient knowledge of compiler development or CS backgrounds should be more responsible for this kind of tasks. In this regard, I always wonder: what is the role of the  current international standard committee here? Are they not consisting of compiler or CS experts and to participate proposals for new features (if they belong to large companies and involved in Fortran related products)? I hope the compiler or CS experts in the committee (if any) will play more active roles in proposing new features to catch up with other modern languages (and make the life easier for Fortran users).

RE `MODULE PROCEDURE NS%FOO`, `EXTERNAL :: NS%FOO`, `VOLATILE NS%FOO` (<-- I understand that the "response" (particularly from noobs) is not needed here, but anyway...), I feel that the first stage for introducing such namespace can be very restrictive, for example, something like "one shall not add new attributes for module entities defined in one scope via qualification by other modules" etc, etc (so as to minimize the workload of considering semantic implications). And if it turns out to be no problem to allow a particular usage later, I guess the latter may be added ("relaxed") later (which seems to be a typical pattern used in Fortran before).

**Comment by septcolor** [comment]
@klauser
> I don't know any of those languages well enough to know whether there's a correspondence or not.

I often refer to Chapel because it is (1) a statically typed language, (2) its syntax was originally influenced by Fortran, (3) its main emphasis is massive parallel computing. However, Chapel is still a very niche language (it is even not within 100th of the TIOBE ranking), and I think it is no wonder that it is not known very well. (And, I am not trying to "sell" Chapel in any way; it has its own limitations that are different from Fortran or other languages.)
https://en.wikipedia.org/wiki/Chapel_(programming_language)

RE Python and Julia, they are already pretty popular in all the language list, but they are dynamically typed languages, so I think there exist inherent differences from statically typed languages (IMO). For example, in Python, module names and other variables are treated on an equal footing, but I guess it is not "equal" in statically typed languages. So, although the comparison with popular languages is useful (in that they increase the degree of familiarity for potential users from other languages), I am afraid there is potential cases that can lead to issues. After all, I also think that experimental implementation would reveal potential issues (based on specifications by experts, if possible...)

**Comment by septcolor** [comment]
RE modules, C++20 now has modules. Other languages I've been learning all use the concept of modules + import now (indeed, things seem to have been converging to the latter). I think the concept of namespaces are of more "low-level" and possibly play more primitive roles as building blocks. However, given that Fortran already has modules, what is the merit of further introducing the concept of namespaces separately? (I am not against it but cannot understand the merit very well because many other languages seem to have no problems with modules only.)

https://en.cppreference.com/w/cpp/language/modules

**Comment by certik** [comment]
I personally feel we don't need namespaces, but I agree with Vipul that we should keep it in mind and not close doors by introducing something that would make it very hard to introduce namespaces later if the community wishes that.

Modules and a well designed "import" is what I personally would like to see. Fortran is very close.

@septcolor regarding who can contribute: I strongly encourage you and anyone else to contribute. The committee and the community is composed of both users and compiler vendors / experts, and we need both to collaborate on new features. Leaving things just to compiler vendors, or just to users to design is not optimal, we need a healthy collaboration of both groups.

**Comment by FortranFan** [comment]
@wclodius2 wrote Oct. 1, 2020 

> There seems to have been a consensus that there should be a poll/vote on the preferred syntax. As there has been a significant pause in postings on this issue, I think it is an appropriate to summarize our current syntactic options to prompt more suggestions for options. There are four orthogonal choices in the syntax:
>
> What word to use as the keyword
> ..

If there are folks who aren't enthused by any of the keywords thus far, here's another one that can perhaps considered for the poll?  **`PREFIX`**:

```Fortran
use, prefix(ut) :: utils, only : savetxt
..
call utXsavetxt(..)
```

**Comment by aradi** [comment]
I think, the idea with `prefix` is a really nice one. But, maybe `prefixed` (as adjective) is more aligned, what we have already, e.g.
```
use, intrinsic, prefixed :: iso_fortran_env
use, intrinsic, prefixed(ife) :: iso_fortran_env
```

**Comment by gronki** [comment]
I don't personally like the prefix idea. It does not feel very modern, and does not introduce anything new to the language (you are able to rename symbols in the current Fortran already). I also feel like it is better to consistently stick to ``%`` as an access operator rather than introducing new ones (just like I do not understand using ``?:`` for inline ``if`` instead of a better syntax). Less is more.

**Comment by bellomia** [comment]
> - use stdlib_stats, as: stats     
> - use stdlib_stats, as is

I would find it rather confusing, not the first one, which I think is neat, but the second: "as is" makes me think "as given" / "raw" / "as provided", so equivalent to the current behavior with`use stdlib_stats` (which I hope you want to preserve (there are cases where one wants a namespace, there are cases where it just adds unnecessary verbosity, that's why python offers also the `from x import z` syntax).

What about 

- `use stdlib_stats, as namespace: stats`
- `use stdlib_stats, as namespace`

instead?

I find it very "natural-english-alike", which is imo what Fortran has always striven for...

**Comment by shahmoradi** [comment]
How about
```
use, intrinsic, namespace :: iso_fortran_env        ! if the entire module name is used as a prefix
use, intrinsic, namespace :: ife => iso_fortran_env  ! if the specified prefix is used
```

**Comment by aradi** [comment]
> use, intrinsic, namespace :: ife => iso_fortran_env  ! if the specified prefix is used

I think, that would be quite Fortranic and nice, I like it a lot! Only, I'd rather go with an adjecive (`prefixed` or `namespaced`) here, as it reads very nicely: "*Use* the *intrinsinc*, *prefixed* module `iso_fortran_env`"

**Comment by bellomia** [comment]
Maybe prefixed, namespaced does not look english

**Comment by ivan-pi** [comment]
Does this usage case help motivate the namespace proposal?

I have a clash between a dummy argument name and a typename. 

```fortran
module object_module
type :: object
end type
contains
subroutine handle_object(object)
class(*), intent(in) :: object
select type(object)  ! the variable
   type is (object)  ! the typename
      ! do something
end select
end subroutine
end module
```

To resolve the name clash, I have the following options
1) Pick a different name for the `object` dummy argument. Since Fortran dummy argument names are part of the public interface, this may not be desirable as it could break existing code.
2) Pick a different name for the derived-type. Also not desirable as it could break existing code.
3) Place the implementation into a submodule, and re-import the type from a separate module, renaming it in the process, e.g.:
```fortran
module object_m
implicit none

type :: object
end type

interface 
module subroutine handle_object(object)
class(*), intent(in) :: object
end subroutine
end interface

end module

module rename_object
use object_m, only: object_m_object => object
public
end module

submodule (object_m) object_impl
contains
   module subroutine handle_object(object)
      use rename_object
      class(*), intent(in) :: object
      select type(object)
         type is (integer)
            print *, object
         type is (object_m_object)
            ! ...
      end select
   end subroutine
end submodule
```

Solution 3 is not very desirable because I suddenly have a new module, just to rename a type. It would be much nicer to just to `object_m % object` or `(object_m) object` (copying the _(module) submodule_ syntax) and refer directly to the entity in the parent module.

**Comment by everythingfunctional** [comment]
> This was put on the feature list for F202Y.
> 
> But then it was dropped in https://j3-fortran.org/doc/year/25/25-119r1.txt.

It was not necessarily dropped, but the initial plan for an implementation was decided to not be ideal. During the discussion it was suggested that maybe a new construct would be a way to go. Something like

```
namespace a
  use b, only: c
end namespace
...
a%c = ...
```

A namespace would then be a first class entity.

So we don't know what the feature will look like yet, but it wasn't dropped from the work list (WG5 has to do that).

**Comment by everythingfunctional** [comment]
> Yes, one “passes” implicitly the module name into nested scopes

Except that currently transitive module names are **not** implicitly brought into scope by a `use` statement. Making that happen not be backwards compatible, because you can't have things with the same name defined in the same scope. I.e. the following example is currently valid, but would become invalid.

```fortran
module a
end module
module b
  use module a
end module
use module b
integer a
end
```

**Comment by everythingfunctional** [comment]
That seems potentially workable, if possibly tricky to define in the standard, but I haven't thought about it real hard yet, and you'll have to convince Data subgroup to revisit it. From your example though, wouldn't `b%something` be more appropriate? 

I'm surprised that example produces an error, and the message doesn't look right. I'm sure you're right that people rarely use the name of a module for something else, but it's still generally better to have backwards compatibility when possible.

**Comment by klausler** [comment]
You can avoid this confusion by omitting the `use` statements entirely.  Instead, use some syntax to distinguish a module name in a qualified identifier, e.g., `%a%something`.  The semantics would be simple -- it's equivalent to having `use a, only: DISTINCTNAME => something` in scope.

**Comment by jacobwilliams** [comment]
I was wondering if any thought has been given to namespaces enabling using libraries in the same project that have the same module names? Currently, I think it's not possible with Fortran? e.g., you have two separate libraries used internally by two separate libraries, but both named `kinds` or something generic like that. Currently, we have to prepend hopefully unique names to modules like this to avoid this. But, maybe something like namespaces would be a better solution here?

**Comment by everythingfunctional** [comment]
> any thought has been given to namespaces enabling using libraries in the same project that have the same module names?

Not really. And even if we did do something to help with that problem, it certainly wouldn't be in a way that doesn't require you to change the module, at which point just renaming the module isn't really that bad.
